### PR TITLE
Implement Slash Command Permissions

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
Implement Slash Command Permissions

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
~~This is blocked by #701~~

### CODE REVIEW

**Comment by FasterSpeeding** [comment]
Add relay methods to Command class

File: hikari/api/rest.py [hunk 1]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
             If an internal error occurs on Discord while handling the request.
         """
 
+    async def fetch_application_guild_commands_permissions(
```
Comment (davfsa) [inline hikari/api/rest.py:6463]: Missing `@abc.abstractmethod` (Same applies for all bellow)

File: hikari/api/rest.py [hunk 2]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
...
+        -------
+        typing.Sequence[hikari.commands.GuildCommandPermissions]
+            Sequence of the guild command permissions set for the specified guild.
+
+        Raises
+        ------
+        hikari.errors.ForbiddenError
+            If you cannot access the provided application's commands or guild.
+        hikari.errors.NotFoundError
+            If the provided application isn't found.
```
Comment (davfsa) [inline hikari/api/rest.py:6487]:
```suggestion
            If the provided application or guild isn't found.
```

File: hikari/api/rest.py [hunk 3]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
...
+        -------
+        hikari.commands.GuildCommandPermissions
+            Object of the command permissions set for the specified command.
+
+        Raises
+        ------
+        hikari.errors.ForbiddenError
+            If you cannot access the provided application's commands or guild.
+        hikari.errors.NotFoundError
+            If the provided application or command isn't found.
```
Comment (davfsa) [inline hikari/api/rest.py:6532]:
```suggestion
            If the provided application, guild or command isn't found.
```

File: hikari/api/rest.py [hunk 4]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
...
+        application: hikari.snowflakes.SnowflakeishOr[hikari.guilds.PartialApplication]
+            Object or ID of the application to set the command permissions for.
+        guild : hikari.undefined.UndefinedOr[hikari.snowflakes.SnowflakeishOr[hikari.guilds.PartialGuild]]
+            Object or ID of the guild to set the command permissions for.
+        permissions : typing.Mapping[hikari.snowflakes.SnowflakeishOr[hikari.commands.Command], typing.Sequence[hikari.commands.CommandPermission]]
+            Mapping of objects and/or IDs of commands to sequences of the commands
+            to set for the specified guild.
+
+            !!! warning
+                Only a maximum of permissions can be set per command.
```
Comment (davfsa) [inline hikari/api/rest.py:6575]: What exactly does this mean?

File: hikari/api/rest.py [hunk 5]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
...
+        -------
+        typing.Sequence[hikari.commands.GuildCommandPermissions]
+            Sequence of the set guild command permissions.
+
+        Raises
+        ------
+        hikari.errors.ForbiddenError
+            If you cannot access the provided application's commands or guild.
+        hikari.errors.NotFoundError
+            If the provided application or command isn't found.
```
Comment (davfsa) [inline hikari/api/rest.py:6587]:
```suggestion
            If the provided application, guild or command isn't found.
```

Comment (davfsa) [reply hikari/api/rest.py:6587]: This also raises the question, if one of them is not found, will it edit until it stumbles on the invalid entity or will it not execute any edits?

File: hikari/api/rest.py [hunk 6]
```diff
@@ -6454,6 +6460,199 @@ async def delete_application_command(
...
+        -------
+        hikari.commands.GuildCommandPermissions
+            Object of the set permissions.
+
+        Raises
+        ------
+        hikari.errors.ForbiddenError
+            If you cannot access the provided application's commands or guild.
+        hikari.errors.NotFoundError
+            If the provided application or command isn't found.
```
Comment (davfsa) [inline hikari/api/rest.py:6638]:
```suggestion
            If the provided application, guild or command isn't found.
```

File: hikari/commands.py [hunk 7]
```diff
@@ -89,7 +96,7 @@ class CommandChoice:
     name: str = attr.field(repr=True)
     """The choice's name (inclusively between 1-100 characters)."""
 
-    value: typing.Union[str, int, float] = attr.field(repr=True)
+    value: typing.Union[str, int] = attr.field(repr=True)
```
Comment (davfsa) [inline hikari/commands.py:99]: Why not float anymore?

File: hikari/api/rest.py (hunk 3 above)
Comment (FasterSpeeding) [reply hikari/api/rest.py:6532]: It's forbidden if the guild isn't found

File: hikari/api/rest.py (hunk 5 above)
Comment (FasterSpeeding) [reply hikari/api/rest.py:6587]: it's the former and when has it ever been the later (also it doesn't edit, it bulk sets)

File: hikari/commands.py [hunk 8]
```diff
@@ -292,3 +306,133 @@ async def delete(self) -> None:
...
+    """Representation of a permission which enables or disables a command for a user or role."""
+
+    id: snowflakes.Snowflake = attr.field(converter=snowflakes.Snowflake)
+    """Id of the role or user this permission changes the permission's state for."""
+
+    type: typing.Union[CommandPermissionType, int] = attr.field(converter=CommandPermissionType)
+    """The entity this permission overrides the command's state for."""
+
+    is_enabled: bool = attr.field()
+    """Whether this permission marks the relevant command as enabled or disabled."""
```
Comment (davfsa) [inline hikari/commands.py:420]: Wouldn't a better name for this be `is_command_enabled`? `is_enabled` sounds like whether the permission is enabled, not the command

**Review by davfsa — COMMENTED** [review]
Everything else looks good :+1:

File: hikari/commands.py (hunk 8 above)
Comment (FasterSpeeding) [reply hikari/commands.py]: It's now called has_access

**Review by davfsa — APPROVED** [review]

**Pull request merged**